File: Problems/Linked-lists/Remove-Nth-Node-From-End-of-List/Remove-Nth-Node-From-End-of-List.cs
using System;

// Using two pointers - Time: O(n)
public class ListNode
{
    public int val;
    public ListNode next;
    public ListNode(int val = 0, ListNode next = null)
    {
        this.val = val;
        this.next = next;
    }
}

public class Solution
{
    public ListNode RemoveNthFromEnd(ListNode head, int n)
    {
        ListNode p = head, q = head;
        while (n > 0)
        {
            q = q.next;
            n--;
        }

        if (q == null)
        {
            return head.next;
        }

        while (q.next != null)
        {
            p = p.next;
            q = q.next;
        }

        p.next = p.next.next;

        return head;
    }
}

class Program
{
    static void PrintList(ListNode head)
    {
        Console.Write("[");
        while (head != null)
        {
            Console.Write(head.val);
            head = head.next;
            if (head != null)
            {
                Console.Write(", ");
            }
        }
        Console.Write("]");
    }
    
    static void Main()
    {
        ListNode head = new ListNode(1, new ListNode(2, new ListNode(3, new ListNode(4, new ListNode(5)))));
        int n = 2;

        Console.Write("Input: head = ");
        PrintList(head);
        Console.WriteLine(", n = {0}", 2);

        Solution sol = new Solution();
        head = sol.RemoveNthFromEnd(head, n);

        Console.Write("Output: ");
        PrintList(head);
        Console.WriteLine();
    }
}

File: Problems/Linked-lists/Remove-Nth-Node-From-End-of-List/Remove-Nth-Node-From-End-of-List.c
#include <stdio.h>
#include <stdlib.h>

// Using two pointers - Time: O(n)

struct ListNode {
    int val;
    struct ListNode *next;
};

struct ListNode *removeNthFromEnd(struct ListNode *head, int n)
{
    struct ListNode *p = head;
    struct ListNode *q = head;
    struct ListNode *prev = NULL;

    while (n > 0)
    {
        q = q->next;
        n--;
    }

    if (q == NULL)
    {
        struct ListNode *newHead = head->next;
        free(head);
        return newHead;
    }

    while (q != NULL)
    {
        prev = p;
        p = p->next;
        q = q->next;
    }

    prev->next = p->next;
    free(p);

    return head;
}

struct ListNode *createList(int values[], int n)
{
    struct ListNode *head = NULL;
    struct ListNode *tail = NULL;

    for (int i = 0; i < n; i++)
    {
        struct ListNode *newNode = (struct ListNode *)malloc(sizeof(struct ListNode));
        newNode->val = values[i];
        newNode->next = NULL;

        if (tail == NULL)
        {
            head = newNode;
            tail = newNode;
        }
        else
        {
            tail->next = newNode;
            tail = newNode;
        }
    }

    return head;
}

void printList(struct ListNode *head)
{
    struct ListNode *current = head;

    printf("[");
    while (current != NULL)
    {
        printf("%d", current->val);
        current = current->next;
        if (current != NULL)
        {
            printf(", ");
        }
    }
    printf("]");
}

int main()
{
    int values[] = {1, 2, 3, 4, 5};
    int n = 2;

    struct ListNode *head = createList(values, sizeof(values) / sizeof(values[0]));

    printf("Input: head = ");
    printList(head);
    printf(", n = 2\n");

    head = removeNthFromEnd(head, n);

    printf("Output: ");
    printList(head);
    printf("\n");

    struct ListNode *current = head;
    while (current != NULL)
    {
        struct ListNode *next = current->next;
        free(current);
        current = next;
    }

    return 0;
}
